### DISCUSSION TITLE
Question on the 'default' console output by MOOSE

### DISCUSSION THREAD

**Discussion by matthiasneuner** [idaholab/moose#16150] [Q&A General] [answered]
Dear Moose-users,

I am currently setting up a Singularity containerized version of MOOSE (and my custom App) for an easy deployment on a HPC system.

However, what I observe is that i obtain a different console output on the HPC compared to my local workstation.
In particular, on the HPC system I get performance timing messages on individual steps:

```
Building mesh ............                                                                 [ 13.78 s]  <---- only on HPC
Initializing equation system ....                                                          [  5.66 s]  <---- only on HPC

Framework Information:
MOOSE Version:           git commit ce49bda782 on 2020-11-06
LibMesh Version:         4f3fa5a6a2104ab8784a6519677589738b9aef6f
PETSc Version:           3.13.3
SLEPc Version:           3.13.3
Current Time:            Wed Nov 11 09:53:02 2020
Executable Timestamp:    Tue Nov 10 20:34:05 2020

Parallelism:
  Num Processors:          24
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 24306
    Local:                 1150
  Elems:                   
    Total:                 3434
    Local:                 146
  Num Subdomains:          2
  Num Partitions:          24
  Partitioner:             metis


Outputting exodus   [  1.11 s]  <---- only on HPC

Time Step 3, time = 0.03, dt = 0.01
Computing initial residual ..                                                              [  3.19 s]  <---- only on HPC
Updating displaced mesh                                                                    [  1.82 s]  <---- only on HPC
 0 Nonlinear |R| = 4.613567e+03
Updating displaced mesh                                                                    [  1.99 s]  <---- only on HPC
  Linear solve converged due to CONVERGED_RTOL iterations 12
Updating displaced mesh                                                                    [  1.76 s]  <---- only on HPC
 1 Nonlinear |R| = 1.897625e+01
Updating displaced mesh                                                                    [  1.83 s]  <---- only on HPC
  Linear solve converged due to CONVERGED_RTOL iterations 18
Updating displaced mesh .                                                                  [  2.05 s]  <---- only on HPC
 2 Nonlinear |R| = 2.414766e+00
Updating displaced mesh                                                                    [  1.87 s]  <---- only on HPC
  Linear solve converged due to CONVERGED_RTOL iterations 14
Updating displaced mesh .                                                                  [  2.84 s]  <---- only on HPC
 3 Nonlinear |R| = 1.142517e-02
Updating displaced mesh .                                                                  [  2.37 s]  <---- only on HPC
  Linear solve converged due to CONVERGED_RTOL iterations 21
Updating displaced mesh                                                                    [  1.65 s]  <---- only on HPC
 4 Nonlinear |R| = 1.064926e-05
Nonlinear solve converged due to CONVERGED_FNORM_RELATIVE iterations 4
 Solve Converged!
```

-- 
I don't obtain these messages on my local workstation. What causes these messages? Of course, the execution (including mpiexec etc.) is completely identical. Since I use for both the HPC system and the local workstation the identical singularity container, there cannot be any differences in the compiled binaries.
Because the performance on the HPC is considerably inferior compared to my workstation, I am currently investigating all peculiarities.

Thank you in advance!

**Comment by aeslaughter** [comment, marked as answer]
Those message only occur when that code in those sections is taking a long time (1s). So, it is just due to the slower system. If you make the problem larger on your workstation you will likely see them appear as well. This feature was actually added for this reason. When you are running on HPC systems they tend to be a bit slower per processor and you also tend to run much larger problems. As such, we were having people think that a simulation was hung, but in truth it was just taking time. These displays give you some indication that your simulation is working.

**Reply by matthiasneuner** [reply]
Perfect, thank you! Considering the size of the problem, the number of processes and the time taken for computing the residual, something must run terribly wrong.

**Comment by ngrilli** [comment]
Dear @matthiasneuner 

We are also considering using singularity image on our HPC.
Could you please share the procedure you used?
Especially for binding the custom app.

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by matthiasneuner** [reply]
Hello, you can find singularity recipes for MOOSE (and third party apps) here in my repo:
https://github.com/matthiasneuner/chamois-singularity

They are intended for my personal MOOSE app 'chamois' https://github.com/matthiasneuner/chamois  (which itself requires a third party lab (Marmot)), but i think you can adapt the singularity recipe files easily for your own MOOSE apps.
Please let me know if you need further support / information.

Best regards, Matthias

**Reply by ngrilli** [reply]
Dear @matthiasneuner 
This is extremely helpful
I will try with my application and let you know how it goes
Best Regards
Nicolò Grilli
National University of Singapore

**Reply by ngrilli** [reply]
Dear @matthiasneuner and @aeslaughter 

I have successfully generated the singularity image for my MOOSE App and it works.

The only problem I found when I transfer the singularity image to our cluster with CentOS
is the "Kernel too old" error.
The kernel in the cluster is version 2.6
but my Ubuntu laptop generates a singularity image with a newer kernel (version > 3).

I wonder if you had a similar problem.
Is there a way out without installing CentOS on a machine I have admin rights?

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by matthiasneuner** [reply]
Hello,

the singularity image itself does not contain a Linux kernel, but rather uses the one provided by the host system.
I never encountered this issue, because our HPC is based on Centos 7 which seems to provide a recent kernel for the Centos 8 and Debian 11 images.
I think you could try to create a singularity image based on an older distribution (e.g., Centos 7 or 6 instead of Centos 8), but I have no experience with those releases and the required changes in order to build MOOSE. Maybe this helps?

Best regards,
Matthias
